### DISCUSSION TITLE
Plugin for rpm-build

### DISCUSSION THREAD

**Discussion by huakim** [rpm-software-management/rpm#3497] [Q&A]
can i write plugin that does extend functionality of rpm-build?

**Comment by pmatilai** [comment]
There are no build-side plugins currently, there hasn't been any need for that. The build-side tends to be things that are best done in macros and scripts rather than C.

What is it that you'd like to do, actually?

**Reply by huakim** [reply]
i wish to write plugin, which will save rpmspec file ast information somewhere, and create api in lua, which can create pre-build, post-build, pre-check, post-check, pre-install, post-install instructions. I think, that plugins can be plugged only if we call special function from rpmlua, with some arguments, but i may be wrong

**Reply by pmatilai** [reply]
No, what I mean is: what do you actually wish to achieve with all that. What is the actual use-case?

**Reply by huakim** [reply]
actually i wish to create lua_generate_subpackages macro, which will generate flavors for lua package, i already did that in https://github.com/huakim/lua-rpm-macros, but i believe i could rewrite this by using new rpm-build api, without parsing spec file, but using already parsed structure, saved somewhere. For example, i could load rpm-build plugin, which does save this structure somewhere and add pre-build functions for creating specparts, parsing that structure. I also believe, that this does prevent macroses to be executed twice or more times, because, while parsing specfile, the macros also will be executed

**Reply by pmatilai** [reply]
Okay so you want to create sub-packages. I still have no idea about what and why, really.

You're better off forgetting dreams about API to save and reuse parsed spec structures, turning that into something exposable as API would be years of work and those years are better spent on something else.

**Reply by huakim** [reply]
But, somehow, when we created tag with name "Source0", the "SOURCE0" and "source0" macros will be created. Maybe, we need invent something to order to work with other tags, for example, with Requires?

**Reply by huakim** [reply]
And, as i noticed, the SOURCE0 macro will not be defined before Source0 tag
